src/bytes.ts

```diff
@@ -0,0 +1,265 @@
+ *
+ * The downside is that the array is created once and then never garbage
+ * collected. This is not a problem in practice because the array is only 256
+ * elements long.
+ *
+ * @returns A function that returns the lookup table.
+ */
+function getPrecomputedHexValuesBuilder(): () => string[] {
+  // To avoid issues with tree shaking, we need to use a function to return the
+  // array. This is because the array is only used in the `bytesToHex` function
+  // and if we were to use a global variable, the array might be removed by the
+  // tree shaker.
+  const lookupTable: string[] = [];
+
+  return () => {
+    if (lookupTable.length === 0) {
+      for (let i = 0; i < 256; i++) {
+        lookupTable.push(i.toString(16).padStart(2, '0'));
+      }
+    }
+
+    return lookupTable;
+  };
+}
+
+/**
+ * Function implementation of the {@link getPrecomputedHexValuesBuilder}
+ * function.
+ */
+const getPrecomputedHexValues = getPrecomputedHexValuesBuilder();
+
+/**
+ * Check if a value is a `Uint8Array`.
+ *
+ * @param value - The value to check.
+ * @returns Whether the value is a `Uint8Array`.
+ */
+export function isBytes(value: unknown): value is Uint8Array {
+  return value instanceof Uint8Array;
+}
+
+/**
+ * Assert that a value is a `Uint8Array`.
+ *
+ * @param value - The value to check.
+ * @throws If the value is not a `Uint8Array`.
+ */
+export function assertIsBytes(value: unknown): asserts value is Uint8Array {
+  assert(isBytes(value), 'Value must be a Uint8Array.');
+}
+
+/**
+ * Convert a `Uint8Array` to a hexadecimal string.
+ *
+ * @param bytes - The bytes to convert to a hexadecimal string.
+ * @returns The hexadecimal string.
+ */
+export function bytesToHex(bytes: Uint8Array): string {
+  assertIsBytes(bytes);
+
+  const lookupTable = getPrecomputedHexValues();
+  const hex = Array.prototype.map.call(bytes, (n) => lookupTable[n]).join('');
+
+  return add0x(hex);
+}
+
+/**
+ * Convert a `Uint8Array` to a `bigint`.
+ *
+ * To convert a `Uint8Array` to a `number` instead, use {@link bytesToNumber}.
+ *
+ * @param bytes - The bytes to convert to a bigint.
+ * @returns The bigint.
+ */
+export function bytesToBigInt(bytes: Uint8Array): bigint {
+  assertIsBytes(bytes);
+
+  const hex = bytesToHex(bytes);
+
+  // This can catch possible regressions in the future, if we ever change the
+  // implementation of `bytesToHex` to return a string without the `0x` prefix.
+  assert(hex.startsWith('0x'), 'Hex string must start with "0x".');
+
+  return BigInt(hex);
+}
+
+/**
+ * Convert a `Uint8Array` to a `number`.
+ *
+ * To convert a `Uint8Array` to a `bigint` instead, use {@link bytesToBigInt}.
+ *
+ * @param bytes - The bytes to convert to a number.
+ * @returns The number.
+ * @throws If the resulting number is not a safe integer.
+ */
+export function bytesToNumber(bytes: Uint8Array): number {
+  assertIsBytes(bytes);
+
+  const number = Number(bytesToBigInt(bytes));
+
+  assert(
+    Number.isSafeInteger(number),
```

Comment:
`Number(BigInt(Number.MAX_SAFE_INTEGER) + 2n) === Number.MAX_SAFE_INTEGER`.

You should check the big int instead.

```
const number = bytesToBigInt(bytes);
assert(number <= BigInt(Number.MAX_SAFE_INTEGER));
return Number(number);
```

Reply:
Didn't know that. Nice catch!